Add debug script

AaronDewes opened the pull request:
If users send the output of `umbrel-debug` with a bug report, it should help to find the problem by providing the information that is often needed. (CPU temperature, Power issues, Network config, swap usage, RPi model, firmware version, Docker logs and the version of Umbrel and Umbrel OS)

AaronDewes commented:
The script now also outputs a short help text that tells the user what they can do to fix their issue.

mayankchhabra reviewed (changes requested):
Thank you so much for this, @AaronDewes! Having a debug script would be extremely helpful. 

Can we please remove the need to install `smartmontools` since we aim to keep the host system as free from additional dependencies as possible?

Also, it'd be great to make this script available on both manual and Umbrel OS installations. So we can move it to `$umbrel/scripts/debug`, check for the existence of `$UMBREL_OS` variable and print additional info (related to throttling, external storage, etc).

A disk space check would be very useful. Couple of users who ran Umbrel on a 500GB drive ran out of space and had to debug their way out of it.

It'd also be helpful to fetch the status/logs of the `umbrel-startup` and `umbrel-external-storage` systemd services on Umbrel OS.

AaronDewes commented on scripts/umbrel-os/debug:
```diff
@@ -0,0 +1,142 @@
+echo "-----------"
+
+cd /home/umbrel/umbrel
+docker-compose logs
```
Added that

AaronDewes commented:
I've fixed the things you mentioned now. The output of `df` already was in filesystem information, but I added -h to make it more readable. I haven't tested the script yet, could you please do that? (Not at home right now, did the edits from my phone)

louneskmt commented on scripts/debug:
```diff
@@ -0,0 +1,144 @@
+# NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS 
+# SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+
+[[ -f "/etc/default/umbrel" ]] && source "/etc/default/umbrel
```
```suggestion
[[ -f "/etc/default/umbrel" ]] && source "/etc/default/umbrel"
```
Missing quotation mark here.

louneskmt commented on scripts/debug:
```diff
@@ -0,0 +1,144 @@
+    echo "----------"
+
+    vcgencmd get_throttled
+
+    echo "Startup status"
```
```suggestion

    echo
    echo "Startup status"
```
Just following the same style

AaronDewes commented:
Thanks, @louneskmt. Applied your suggestions.

mayankchhabra commented on scripts/debug:
```diff
@@ -0,0 +1,146 @@
+echo "Docker logs"
+echo "-----------"
+
+docker-compose logs --tail=100
```
On second thought, it might be a better idea to individually fetch logs of specific containers because one container could disproportionately spam the entire output of `docker-compose logs`. 

Plus we haven't really ever had the need to look at the logs of middleware, manager, nginx and dashboard containers so we can just rely on their statuses.

```suggestion
docker ps --format="table {{.Names}}\t{{.Status}}"

echo
echo "Bitcoin Core logs"
echo "-----------"
echo
docker-compose logs --tail=30 bitcoin

echo
echo "LND logs"
echo "-----------"
echo
docker-compose logs --tail=30 lnd

echo
echo "Tor logs"
echo "-----------"
echo
docker-compose logs --tail=30 tor
```

mayankchhabra commented on scripts/debug:
```diff
@@ -0,0 +1,146 @@
+        exit 0
+    fi
+fi
+echo "The script could not automatically detect an issue with your Raspberry Pi."
```
```suggestion
echo "This script could not automatically detect an issue with your Umbrel."
```

mayankchhabra commented on scripts/debug:
```diff
@@ -0,0 +1,146 @@
+    fi
+fi
+echo "The script could not automatically detect an issue with your Raspberry Pi."
+echo "Please share the output of this command in the Umbrel Telegram group, and we will help you resolve the problem."
```
```suggestion
echo "Please copy the entire output of this script and paste it in the Umbrel Telegram group (https://t.me/getumbrel) so we can help you with your problem."
```

mayankchhabra commented on scripts/debug:
```diff
@@ -0,0 +1,146 @@
+ifconfig | sed -e "s/\([0-9a-fA-F]\{1,4\}:\)\{7,7\}[0-9a-fA-F]\{1,4\}/y.y.y.y.y.y.y.y/g" | sed -e "s/[0-9a-fA-F]\{1,4\}:\(:[0-9a-fA-F]\{1,4\}\)\{1,4\}/y::y.y.y.y/g" | sed -e "s/\([0-9]\{1,3\}\.\)\{3,3\}[0-9]\{1,3\}/x.x.x.x/g" | sed -e "s/\([0-9a-fA-F]\{2,2\}\:\)\{5,5\}[0-9a-fA-F]\{2,2\}/m.m.m.m/g"
+
+echo
+echo "Docker logs"
```
```suggestion
echo "Docker containers"
```

mayankchhabra commented on scripts/debug:
```diff
@@ -0,0 +1,146 @@
+    vcgencmd get_throttled
+
+    echo
+    echo "Startup status"
```
```suggestion
    echo "umbrel-startup.service logs"
```

mayankchhabra commented on scripts/debug:
```diff
@@ -0,0 +1,146 @@
+    echo
+    echo "Startup status"
+    echo "--------------"
+    systemctl status umbrel-startup
```
Since `systemctl` pages the logs, it'd be easier to fetch the last 30 lines via `journalctl`:

```suggestion
    journalctl --unit=umbrel-startup.service | tail -n 30
```

mayankchhabra commented on scripts/debug:
```diff
@@ -0,0 +1,146 @@
+
+docker-compose logs --tail=100
+
+if [[ ! -z "${UMBREL_OS:-}" ]]; then
```
It's safe to remove this entire `if` block since `umbrel-external-storage.service` already performs these sanity checks for no drive/more than 1 drive and we can spot any errors in its logs.

AaronDewes replied:
I added this because it then can give the output of the check and directly tell the user the problem so they don't have to ask.  (Lines 140 - 152)

mayankchhabra replied:
Ah, okay. Yeah that makes sense!

mayankchhabra commented on scripts/debug:
```diff
@@ -0,0 +1,146 @@
+# SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+# This fork is Copyright: 2020 Umbrel.
+
+[[ -f "/etc/default/umbrel" ]] && source "/etc/default/umbrel"
```
We can drop this since the file is already [sourced in `/etc/profile`](https://github.com/getumbrel/umbrel-os/blob/master/stage2/04-set-umbrel-os-version/00-patches/01-profile.diff#L11).

```suggestion
```

AaronDewes replied:
@mayankchhabra Can this then also be removed [in the update scripts](https://github.com/getumbrel/umbrel/blob/master/scripts/update/01-run.sh#L20)?

mayankchhabra replied:
Update scripts run as sudo and `/etc/profile` isn't automatically sourced for sudo :(

AaronDewes replied:
@mayankchhabra Both users that used the script in the telegram group yet have had the same issue: the script didn't detect Umbrel OS. Should I add it again?

AaronDewes replied:
Maybe you need to add an "export " to the file?

mayankchhabra replied:
Good spot, Aaron! That is exactly it.

Let's go with sourcing the file for now and then eventually we'll update the `/etc/default/umbrel` file with an export in an OTA update.

mayankchhabra replied:
Good spot, Aaron! That is exactly it.

Let's go with sourcing the file for now and then eventually we'll update the /etc/default/umbrel file with an export in an OTA update.

mayankchhabra reviewed (approved):
Great job, Aaron! You've reduced tonnes of painful debugging to just a one-liner `ssh -t umbrel@umbrel.local "/home/umbrel/umbrel/scripts/debug"` that we can now pass to anyone facing (almost) any kind of issue with their Umbrel! Awesome work. 👏

Let's for @lukechilds to have a quick look at the script as well before we merge.

AaronDewes commented:
A few more improvements are available now, thanks to @louneskmt  for suggesting them.

- The script can now be run with `--upload` to automatically generate a link to share for debugging
- .onion URLs are now no longer in the output
- A few other minor improvements to the result text

mayankchhabra commented on scripts/debug:
```diff
@@ -0,0 +1,183 @@
+echo
+cat /proc/swaps
+
+
+echo
+echo "Networking information"
+echo "----------------------"
+echo
+
+ifconfig | sed -e "s/\([0-9a-fA-F]\{1,4\}:\)\{7,7\}[0-9a-fA-F]\{1,4\}/y.y.y.y.y.y.y.y/g" | sed -e "s/[0-9a-fA-F]\{1,4\}:\(:[0-9a-fA-F]\{1,4\}\)\{1,4\}/y::y.y.y.y/g" | sed -e "s/\([0-9]\{1,3\}\.\)\{3,3\}[0-9]\{1,3\}/x.x.x.x/g" | sed -e "s/\([0-9a-fA-F]\{2,2\}\:\)\{5,5\}[0-9a-fA-F]\{2,2\}/m.m.m.m/g"
+
```
We can probably get rid of the networking info as it takes too much space in the logs and we never actually had the need to look at it and instead let's tail `karen's` logs as those are quite useful in the event of a failed update, backup, app installs/uninstalls, etc.

```
tail -n 50 logs/karen.log | sed '/onion/d'
```

mayankchhabra reviewed (changes requested):
Really awesome improvements, Aaron! Let's get this in the next release.

AaronDewes commented:
@mayankchhabra Done. I also added the dmesg output (upload only) because that can help research SSD issues.

lukechilds reviewed (approved):
Awesome stuff @AaronDewes, sorry for the delay in reviewing this!
